Introduced pattern-based simplification for Boolean functions with Relational

oscargus opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234". See
https://github.com/blog/1506-closing-issues-via-pull-requests . Please also
write a comment on that issue linking back to this pull request once it is
open. -->

#### Brief description of what is fixed or changed
Added support for pattern-based simplification of `BooleanFunction`s with `Relational`s.

For example, `And(x >= y, Eq(x, y)).simplify()`gives `Eq(x, y)`

Combined with #15905 (and to some extend #15898) this will significantly improve the simplification capability of these types of expressions. When #15899 is solved such that there is a consistent canonical form, the current code can be simplified to only consider the cases that can happen based on the canonical form.

#### Other comments


#### Release Notes

<!-- Write the release notes for this release below. See
https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more information
on how to write release notes. The bot will check your release notes
automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
* logic
   * Improved simplification of Boolean functions with relations in them
<!-- END RELEASE NOTES -->

oscargus committed:
Improved simplification of logic expressions with relationals

oscargus committed:
Extracted a common method for pattern-based simplification

oscargus committed:
Improved comments for the unified version

oscargus committed:
Fixed small bug

smichr commented:
cf #8373 and discussion in linked issues/PRs

oscargus commented:
Thanks @smichr ! I thought about the real vs non-real issue and in the original code in #15905 I had a check that variables were not simplified away in case it later turned out that they were complex (see the first commits). However, I noted that some of the original parsing ignored this, so I also ignored it. The same thing happens here, although it can be rather straightforwardly introduced by only allowing the simplifications to happen on symbols/expressions where `is_real` is True. (Note that `Lt` etc throws an exception for complex variables, hence, `x**2 < 0` cannot, strictly, be evaluated for complex numbers, at least according to my interpretation.)

Keeping +/-oo out of the relations in not considered (that is more of a #15905 issue though).

I'll think a bit more about it, but I am quite sure that this is "safer" than #15905.

One feature of Maple which I think is quite nice is "simplify symbolic", which basically skips the checks that everything is real, finite etc, and just simplifies. I do not know if there is a switch for sympy that gives the same thing, but that may be something to consider. Of course it is very useful that sympy does not introduce incorrect mathematical conclusions, but quite often (for me at least), the problems have the characteristics that this isn't really a problem and rather than specifying properties of all symbols etc, one just would like to use the CAS to obtain a simplified solution without the risk of typos etc associated with paper and pen work (and with a bit more power/knowledge when it comes to simplifications).

smichr commented:
Feel free to use anything you see in the closed PRs.

smichr commented on sympy/logic/tests/test_boolalg.py:
```diff
@@ -828,3 +829,14 @@ def test_binary_symbols():
+
+
+def test_relational_simplification():
+    assert Or(x >= y, x < y).simplify() == S.true
```
This can only be True if x and y are comparable. If either or both of x and y are imaginary then this cannot be evaluated.

oscargus replied:
Yes, that is an inherent problem with all of this. The question is just if one should assume anything put in a comparison to be non-complex (I think there will be an exception if it is explicitly complex) or force the user to explicitly declare everything as real and check that for all simplifications.

I envision that x**2 < 0 may come up as an example where complex is valid, but I honestly think that x < I + 1 is as valid. In both cases the solutions are on a single line in the complex plane. We do not allow the second, so what is so special about the first?

smichr replied:
The first is comparing two real numbers when x is pure imaginary.

oscargus replied:
The second also compares two real numbers, but along a different line in the complex space.

oscargus replied:
I think the main argument though is that since these type of relations in general are not valid for complex numbers, it makes sense to assume that the inputs are real-valued.

smichr replied:
> second also compares two real numbers

Well, `1 + I` is not a real number, but yes, we can define it so that comparison of complex numbers is done by comparing the real parts. But that is not something that we allow (at present), choosing, rather, to raise an error for `1 < I` rather than False. I am not so savvy on all the implications. We try to be careful about assumptions and I don't see how we can, without user permission, say something is True or False when there are implicit assumptions being made. We have methods like `posify` and options like `force` in place. Perhaps simplify could get a flag `real=True` that would assume that vanilla symbols are real. (Actually, a more general `vanilla=dict(real=True, positive=...)` might be better where anything that is vanilla takes on the attributes passed so `vanilla=dict(postive=True)` could be used to make all vanilla symbols positive.)

oscargus replied:
Yes, but when x is pure imaginary it is not a real number either.

I am not sure if we are fully understanding each other. My main argument is really that relations should imply real-value numbers as there is not a valid way to compare otherwise (except for Eq and Ne). Hence, making that assumption when simplifying should be OK. Right now it seems like one can sometimes use complex numbers, `x**2 < 0` implies a complex number (`Re(x) == 0` and `Im(x) != 0` for it to not be False) and is OK (since it not simplifies to False), `x < I + 1` (or more clearly, `x - I < 1` implies a complex number (`Im(x) == 1` leads to a real-valued comparison) and is not OK.

Maybe the way to go is to implement `relsimp` where it is clearly stated that it is assumed that the simplifications are valid for real-valued symbols only and that oo and -oo may be incorrectly handled (the `x(x+1) > x**2+x` case which in my approach will be False, not sure if this is the exactly correct formulation of where it is an issue). There is also the #15905 PR that can go into this function that has related assumptions. The only drawback I can see is that the code requires a bit of wrappers around it and must be written in a different way (or maybe one just renames* the current `_eval_simplify` to `_eval_relsimp` and writes a simple `relsimp` based on `simplify`).

* Extract the added parts to `_eval_relsimp` and keeps the previous ones.

smichr replied:
> Yes, but when x is pure imaginary it is not a real number either.
> 
> I am not sure if we are fully understanding each other

What I mean is that `x**2 < 0` becomes `-1 < 0` when x = I and that is a perfectly fine comparison, likewise for `x - I < 0` which becomes `0 < 0`. Both are legitimate comparisons.

> Maybe the way to go is to implement `relsimp`

My initial thought is that we already know how to deal with this in a sympythonic way: make the variables real. If someone is working with real variables they can jsut do `vars('a:z', real=True)` or if we make a `flavor` function then `flavor(real=True)` would change all vanilla symbols to real.

oscargus replied:
The thing is that `x - I < 1` will throw an exception on you at the moment. So for consistency reasons `x**2 < 0` might just as well simplify to False. The obvious option is of course to not throw an exception for `x - I < 1`.

Of course a real check is an option. However, that require additional checking of every variable, which is maybe not a big deal though compared to the matching. I can try to include those both here and in the other PR.

oscargus replied:
@smichr Try the following in your current sympy:

```
from sympy import symbols, Or
x, y = symbols('x y')
Or(x >= y, x < y)
```

smichr replied:
Hmm...it's already True without simplify. So it looks like we have a precedent to assume that vanilla symbols are real for purposes of comparison.

> that `x - I < 1` will throw an exception on you at the moment
In your PR? In master it doesn't throw an error.

oscargus replied:
I was wrong there. `x - 1 < I` will though. In #15905 I have added a piece of code that will keep the imaginary constants in the exact position of the relation, so they won't simplify away in e.g. `x - I < y - I`. Not sure if I've pushed it though (doesn't look like it). Not sure if one can do something similar here, although I do not think it is an issue in the same way here.

asmeurer commented:
This is a nice approach. Do the patterns match subexpressions as well, like if an And or Or has more than 2 arguments?

Can we add some automatic testing of the patterns. I think what we can do is take the three variables and plug in the values 0, 1, 2 for them, then check that the lhs reduces to the same boolean as the rhs. Actually we can skip certain combinations, as a=b=c=0 will be logically the same as a=b=c=1, and the same with a=b=0,c=1 and a=b=1,c=2 (basically, you can skip combinations that don't have 0). 

Most Piecewise from SymPy come from integrate, which tends to produce similar looking conditions (they usually come from the meijerg convergence conditions), so pattern based simplify is a good approach. We will need some more advanced patterns to really handle the sorts of things returned by integrate, but this is a good start.

oscargus commented:
@asmeurer I do not know, but clearly a relevant question. I will check the next time I give it a go.

oscargus commented:
11 days is a long time apparently. Yes, they do.

Basically I test all combinations pairwise, replace the best simplification and tests all combinations again. This can of course be simplified to not retest combinations that are already tested and simply test the new possible combinations, but that is not done yet.

The testing for this is limited as well.

oscargus committed:
Simplification only for explicitly real variables

oscargus committed:
Do not just run the tests, check the results...

oscargus commented:
OK, so I've updated this to only simplify expressions where all free symbols is_real. It seems to work (see the last test in the file, only the x and y expression simplifies, not the d and e one).

asmeurer commented:
How about allowing is_real = None and only disallowing is_real = False.

asmeurer commented:
Also I think the testing of the patterns with 0, 1, 2 should be in the test suite, if that wasn't clear.

oscargus commented:
There is also `is_comparable` which I have still to figure out how it related to the other properties.If I get it correctly it may not be the solution though (although the name somehow indicates it), as it from a quick glance appears as it means "can be evaluated to an actual number".

I'd be happy with the `is_real in (True, None)`-approach. Somehow, I still think that, yes, there will be cases where a complex may slip through, but writing a comparison when something can be complex is (most of the time) incorrect anyway.

I'm not sure if I got the 0, 1, 2 approach fully (and I forgot about it yesterday either way), but I will think about it and give it a go from my interpretation. So basically, have a bunch of rather long expressions with logic and relations with a number of variables, apply simplifications, subs values into the non-simplified and the simplified expressions and see that they evaluate to the same boolean value?

oscargus committed:
Added a more general test with numbers

oscargus commented:
Now I added a test that for n free symbols tries all combinations of values between -(n-1) and n-1 for all symbols. This clearly allows testing arbitrary large expressions where one not necessarily knows if the result simplification is correct.

I also added that in case of an error, the exact expressions and combinations are printed to simplify debugging (if it now ever should happen...).

asmeurer commented on sympy/logic/tests/test_boolalg.py:
```diff
@@ -843,3 +844,31 @@ def test_relational_simplification():
+            if originalvalue != simplifiedvalue:
+                # In case something fails, show clearly which case
+                print("Original: {}\nand simplified: {}\ndo not evaluate to the same value for {}".format(original, simplified, sublist))
+            assert originalvalue == simplifiedvalue
```
You can add this to the end of the assert, like `assert originalvalue == simplifiedvalue, "Original: {}\nand simplified: {}\ndo not evaluate to the same value for {}".format(original, simplified, sublist)`. That will make it the error message of the AssertionError.

oscargus replied:
Thanks!

asmeurer commented:
I meant to test the patterns themselves. That is, numerically test the result of `simplify_patterns_or` and `simplify_patterns_and`. If the patterns are correct, then we only need to trust the pattern substitutor.

smichr commented on sympy/logic/boolalg.py:
```diff
@@ -498,6 +498,107 @@ def _eval_derivative(self, x):
+                    # Insert the new one (no need to insert a value that will
+                    # not affect the result)
+                    Rel.append(newrel.canonical)
+                # We did change comething so try again
```
comething->something

oscargus replied:
Thanks! Will fix. (I'll wait until the improved canonical is hopefully merged as it will simplify this quite a bit.)

oscargus commented:
@asmeurer Ahh, makes sense! Will add that. (May wait with pushing until #16004 is hopefully merged as this will lead to significant simplifications of this PR.)

oscargus committed:
Merge branch 'master' into simplifybooleanwithrelational

oscargus committed:
Fixed tests and added simplification for Xor

oscargus commented:
This is now updated with better testing based on the suggestions from @asmeurer  and also patterns for Xor. The testing helped me find a few issues so clearly that was a good idea.

I also changed it so that it skips the simplifications only if `is_real == False`.

What are good ways to get expressions with `Piecewise`? Ideally I would like to find "naturally" occurring cases of Boolean expressions with `Relational`s to try it out and possibly find more useful patterns (the testing also relates to the `Piecewise` simplification in #15929 ).

I will need to rethink the redundancy of tests here. Even though we can always rewrite the `Relational` so that they do not include, e. g. `Ge`and `Gt`, that will introduce issues with the signs and one will need to add more patterns with different signs. Currently, I am quite sure that half of the min/max patterns are not required, as the current approach with reversing the `Relational`s and test all four combinations will lead to one of the other ones (so `Ge` `Ge` will also show up as `Le` `Le`). Apart from that it is not really clear...

smichr commented:
> What are good ways to get expressions with `Piecewise`

What do you mean?

oscargus commented:
I want to get "naturally" occurring expressions using Piecewise to test this out with. Typically of the kind that an ordinary user would get (and, hence, would be good candidates to see if they simplify efficiently. It is one thing to generate synthetic cases, but there may be things I do not think about.

I can of course try to run solveset, integrate and piecewise_fold with random expressions, but I was thinking that maybe someone had some good examples of cases where a Piecewise (or a Boolean expression with Relationals) is obtained that can be simplified.

oscarbenjamin commented:
This PR looks good.

I see references to things being able to be simplified with better canonical. There have been many canonical PRs. Are they merged now?

oscargus commented:
Yes, "my" canonical PR is merged now. However, I have some doubt regarding how much actually can be simplified as a consequence of it.

There is probably some redundant checks in it right now, but I still think that this can be merged and continued later.
